recipe-ts-runtime/src/Recipe.ts

```diff
@@ -1,4 +1,139 @@
 import { Ingredient } from "./Ingredient";
+import { KeyedIngredient } from "./KeyedIngredient";
+
+export interface Segment {
+    domain: string | null;
+    recipe: Recipe;
+}
 
 export class Recipe extends Ingredient {
+    private readonly ingredients: Ingredient[];
+    private context: string | null = null;
+    private contextIngredient: KeyedIngredient | null = null;
+
+    public static prepare(...ingredients: Ingredient[]): Recipe {
+        return new Recipe(...ingredients);
+    }
+
+    public static context(context: string | KeyedIngredient, ...ingredients: Ingredient[]): Recipe {
+        const recipe = new Recipe(...ingredients);
+        recipe.context = (typeof context === "string") ? context : null;
+        recipe.contextIngredient = (context instanceof KeyedIngredient) ? context : null;
+        return recipe;
+    }
+
+    public getIngredients(): Ingredient[] {
+        return this.ingredients;
+    }
+
+    public getContext(): string | null {
+        return this.context;
+    }
+
+    public getContextIngredient(): KeyedIngredient | null {
+        return this.contextIngredient;
+    }
+
+    public segment(): Segment[] {
+        const segments: Segment[] = [];
+        const recipeStack: Recipe[] = [];
+        recipeStack.push(this);
+        this._segment(new Recipe(), recipeStack, null, segments);
+        return segments;
+    }
+
+    private _segment(currRecipe: Recipe | null, recipeStack: Recipe[], currDomain: string | null, segments: Segment[]) {
+        if (recipeStack[0].contextIngredient !== null) {
+            const contextIngredient = recipeStack[0].contextIngredient;
+            if (contextIngredient!.getDomain() !== currDomain) {
```

Comment:
Can probably put the `!` on the previous line to avoid using it everywhere `contextIngredient` is used

Reply:
Weird that the first `if` doesn't prevent it actually. I assume all these `!`s after `contextIngredient` are because `contextIngredient` has type `X | null`? I thought typescript would be smart enough to realize we checked that it's not null on the first line... Putting `recipeStack[0].contextIngredient` in a variable might also fix it.

Reply:
Do you mean put the `!` on the `const contextIngredient = ...`? That doesn't work. Unfortunately, TypeScript isn't smart enough to look at the above condition...

Reply:
Something seems wrong then... what is the type of `contextIngredient`? I'm assuming it is `Ingredient | null`, and I think we should somehow cast it to `Ingredient` once in the declaration instead of once per use. Could either use `!` or `as Ingredient` or `<Ingredient>` etc.

But also... normally if you do a null-check in an `if` then TS is smart enough to realize that variable will never be null, but maybe that only works for simple vars and not for things like `a[0].b`.

None of this actually change the compiled JS, but imo makes the code nicer.

Reply:
It's possible that the null check is getting throw away because we're accessing `contextIngredient` through `recipeStack[0].contextIngredient` and typescript doesn't want to assume that indexing by a number always returns the same result. I'll try putting it in a variable.

Reply:
Yeah, that fixed it.